bee-crypto/src/ternary/sponge/batched_curlp/bct.rs

```diff
@@ -9,95 +9,119 @@
 // an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 // See the License for the specific language governing permissions and limitations under the License.
 
-use std::slice::SliceIndex;
+use std::ops::{Deref, DerefMut, Range};
+
+#[derive(Clone, Copy, Debug)]
+pub struct BCTrit(pub usize, pub usize);
+
+impl BCTrit {
+    const fn zero() -> Self {
+        Self(0, 0)
+    }
+
+    pub fn lo(&self) -> usize {
+        self.0
+    }
+
+    pub fn hi(&self) -> usize {
+        self.1
+    }
+}
 
 #[derive(Clone)]
 pub struct BCTritBuf {
-    lo: Vec<usize>,
-    hi: Vec<usize>,
+    inner: Vec<BCTrit>,
 }
 
 impl BCTritBuf {
-    pub fn as_slice(&self) -> BCTritRef<'_, [usize]> {
-        BCTritRef {
-            lo: &self.lo,
-            hi: &self.hi,
+    pub fn zeros(len: usize) -> Self {
+        Self {
+            inner: vec![BCTrit::zero(); len],
         }
     }
 
-    pub fn as_slice_mut(&mut self) -> BCTritMut<'_, [usize]> {
-        BCTritMut {
-            lo: &mut self.lo,
-            hi: &mut self.hi,
+    pub fn filled(value: usize, len: usize) -> Self {
+        Self {
+            inner: vec![BCTrit(value, value); len],
         }
     }
+}
+
+impl Deref for BCTritBuf {
+    type Target = BCTrits;
+
+    fn deref(&self) -> &Self::Target {
+        unsafe { &*(self.inner.deref() as *const [BCTrit] as *const BCTrits) }
+    }
+}
+
+impl DerefMut for BCTritBuf {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        unsafe { &mut *(self.inner.deref_mut() as *mut [BCTrit] as *mut BCTrits) }
+    }
+}
 
+#[repr(transparent)]
+pub struct BCTrits {
+    inner: [BCTrit],
+}
+
+impl BCTrits {
     pub fn fill(&mut self, value: usize) {
-        for (lo, hi) in self.lo.iter_mut().zip(self.hi.iter_mut()) {
+        for BCTrit(hi, lo) in &mut self.inner {
             *lo = value;
             *hi = value;
         }
     }
 
-    pub fn filled(value: usize, len: usize) -> Self {
-        BCTritBuf {
-            lo: vec![value; len],
-            hi: vec![value; len],
-        }
+    pub fn len(&self) -> usize {
+        self.inner.len()
     }
 
-    pub fn zeros(len: usize) -> Self {
-        Self::filled(0, len)
+    pub fn copy_from_slice(&mut self, slice: &Self) {
+        self.inner.copy_from_slice(&slice.inner)
     }
 
-    pub fn len(&self) -> usize {
-        self.lo.len()
+    pub unsafe fn get_unchecked<I: BCTritsIndex>(&self, index: I) -> &I::Output {
+        index.get_unchecked(self)
     }
 
-    pub fn lo(&self) -> &[usize] {
-        &self.lo
+    pub unsafe fn get_unchecked_mut<I: BCTritsIndex>(&mut self, index: I) -> &mut I::Output {
+        index.get_unchecked_mut(self)
```

Comment:
Why is the get applied on the index ?

Reply:
It's inspired by https://doc.rust-lang.org/std/slice/trait.SliceIndex.html 

Basically it's because each index type (`usize` and `std::ops::Range<usize>`) implements their own specific `get_unchecked_mut` with its own output so `BCTrits::get_unchecked` can be sort of "overloaded" to work with integers and ranges.

Reply:
Alright then, first time I see this pattern